src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.cpp

```diff
@@ -0,0 +1,340 @@
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+
+// Compute the VolumeDim-dimensional quadrature weights, by taking the tensor
+// product of the 1D quadrature weights in each logical dimension of the Mesh.
+template <size_t VolumeDim>
+DataVector volume_quadrature_weights(const Mesh<VolumeDim>& mesh) noexcept {
+  std::array<DataVector, VolumeDim> quadrature_weights_1d{};
+  for (size_t d = 0; d < VolumeDim; ++d) {
+    gsl::at(quadrature_weights_1d, d) =
+        Spectral::quadrature_weights(mesh.slice_through(d));
+  }
+
+  DataVector result(mesh.number_of_grid_points(), 1.);
+  for (IndexIterator<VolumeDim> s(mesh.extents()); s; ++s) {
+    result[s.collapsed_index()] = quadrature_weights_1d[0][s()[0]];
+    for (size_t d = 1; d < VolumeDim; ++d) {
+      result[s.collapsed_index()] *= gsl::at(quadrature_weights_1d, d)[s()[d]];
+    }
+  }
+  return result;
+}
+
+// Compute the matrix that interpolates data in VolumeDim dimensions from
+// source_mesh to the grid points of target_mesh, by taking the tensor product
+// of the 1D interpolation matrices in each logical dimension of the meshes.
+//
+// Note: typically it is more efficient to interpolate from source_mesh to
+// target_mesh in one logical direction at a time, and to avoid computing this
+// matrix. However, the HWENO algorithm requires constructing this matrix
+// explicitly.
+//
+// In particular, for this HWENO application:
+// - source_mesh is the mesh of local element.
+// - target_mesh is the mesh of the neighboring element in direction. Note that
+//   there can only be one neighbor in this direction, and this is checked in
+//   the call to `neighbor_grid_points_in_local_logical_coordinates`.
+// - rectilinear elements are assumed.
+template <size_t VolumeDim>
+Matrix volume_interpolation_matrix(
+    const Mesh<VolumeDim>& source_mesh, const Mesh<VolumeDim>& target_mesh,
+    const Element<VolumeDim>& element,
+    const Direction<VolumeDim>& direction) noexcept {
+  // The grid points of source_mesh and target_mesh must be in the same
+  // coordinates to construct the interpolation matrix. Here we get the points
+  // of target_mesh in the local logical coordinates.
+  const auto target_1d_coords =
+      SlopeLimiters::Weno_detail::neighbor_grid_points_in_local_logical_coords(
+          source_mesh, target_mesh, element, direction);
+
+  const auto interpolation_matrices_1d =
+      intrp::RegularGrid<VolumeDim>(source_mesh, target_mesh, target_1d_coords)
+          .interpolation_matrices();
+
+  // The 1D interpolation matrices will be empty if there is no need to
+  // interpolate in that particular direction (i.e., the interpolation in that
+  // direction is identity). This function undoes the optimization by returning
+  // elements of an identity matrix if an empty matrix is found.
+  const auto matrix_element = [&interpolation_matrices_1d](
```

Comment:
That's right, the constrained fit scales as n^6, as does the measurement of the oscillation indicator (merged in an earlier PR). You should be sad... but mostly because of how poorly we understand limiting 😛